parser-typechecker/src/U/Codebase/Branch/Diff.hs

```diff
@@ -113,34 +117,37 @@ instance Semigroup NameBasedDiff where
     NameBasedDiff (terms0 <> terms1) (types0 <> types1)
 
 -- | Diff two Branches, returning a tree containing all of the changes
-diffBranches :: forall m. (Monad m) => Branch m -> Branch m -> m TreeDiff
+diffBranches :: forall m. (Monad m) => Branch m -> Branch m -> m (TreeDiff m)
 diffBranches from to = do
   let termDiffs = diffMap (Branch.terms from) (Branch.terms to)
   let typeDiffs = diffMap (Branch.types from) (Branch.types to)
   let defDiff = DefinitionDiffs {termDiffs, typeDiffs}
-  childDiff <- do
-    Align.align (children from) (children to)
-      & wither \case
-        This ca -> do
-          -- TODO: For the names index we really don't need to know which exact
-          -- names were removed, we just need to delete from the index using a
-          -- prefix query, this would be faster than crawling to get all the deletes.
-          removedChildBranch <- Causal.value ca
-          Just . unTreeDiff <$> diffBranches removedChildBranch Branch.empty
-        That ca -> do
-          newChildBranch <- Causal.value ca
-          Just . unTreeDiff <$> diffBranches Branch.empty newChildBranch
-        These fromC toC
-          | Causal.valueHash fromC == Causal.valueHash toC -> do
-              -- This child didn't change.
-              pure Nothing
-          | otherwise -> do
-              fromChildBranch <- Causal.value fromC
-              toChildBranch <- Causal.value toC
-              diffBranches fromChildBranch toChildBranch >>= \case
-                Lens.Empty -> pure Nothing
-                TreeDiff cfr -> pure . Just $ cfr
-  pure $ TreeDiff (defDiff :< childDiff)
+  let childDiff = do
+        Align.align (children from) (children to)
+          & wither \case
+            This ca -> do
+              -- TODO: For the names index we really don't need to know which exact
+              -- names were removed, we just need to delete from the index using a
+              -- prefix query, this would be faster than crawling to get all the deletes.
+              removedChildBranch <- Causal.value ca
+              Just . unTreeDiff <$> diffBranches removedChildBranch Branch.empty
+            That ca -> do
+              newChildBranch <- Causal.value ca
+              Just . unTreeDiff <$> diffBranches Branch.empty newChildBranch
+            These fromC toC
+              | Causal.valueHash fromC == Causal.valueHash toC -> do
+                  -- This child didn't change.
+                  pure Nothing
+              | otherwise -> do
+                  fromChildBranch <- Causal.value fromC
+                  toChildBranch <- Causal.value toC
+                  diffBranches fromChildBranch toChildBranch >>= \case
+                    TreeDiff (defDiffs :< Compose mchildren) -> do
+                      children <- mchildren
```

Comment:
I agree with this comment :thinking:

Reply:
Strange, in practice it seems to work just fine, but I agree this looks like an error. I think it might just not be triggering this case often enough to affect the memory footprint, but at any rate I'll switch it around and see if it works out 👍🏼 

This was meant to be an optimization so that you don't see child trees that don't have a diff in them, but I think as it is it's just a bug, I'll swap that around 😄

Reply:
Ok. And is it still clear to you, despite this bug, why we see memory improvements? I had thought that calculating a diff one layer at a time was the source of the streaming.

Reply:
Okay I dug into this a bit and figured it out. The use-case which I cared the most about was building a name lookup from scratch, since that's the use-case that uses the most memory. But in that situation we're diffing against the empty branch, so we never hit this case where we have to recursively diff children at the same position. 
The problematic case would only be triggered on subsequent diffs, and usually in those situations the diff is small enough that memory isn't as much of a concern; but if someone pushed a very large diff that shared a spine with the previous branch then it would trigger this and use more memory than needed.

So TLDR; this problem would've shown up as a bit of extra memory usage in subsequent pushes where the from and to branches were close to each other.

Great catch on this one @tstat!